### PR TITLE
aws_request_signing: a few fixes

### PR DESCRIPTION
**Pull request by rgs1**
There were a few things missing to make this filter work with S3:

* empty body requests (e.g.: GET) still need a sha256 content header
  (the empty string hash)
* requests for s3 shouldn't sign all headers (especially because some
  of them might change, e.g.: x-forwarded-for)
* when proxying requests to s3, HCM's host rewrite won't work so
  support it as a filter config option
* ditto for prefix rewrite [ not supported yet ] 

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
cc: @lavignes @marcomagdy

**Comment by marcomagdy** [comment]
seems like some changes overlap with the [Lambda PR](https://github.com/envoyproxy/envoy/pull/10260) I submitted yesterday...

**Comment by dio** [comment]
@rgs1 "coverage" is currently not working.

File: source/extensions/common/aws/utility.cc [hunk 1]
```diff
@@ -25,19 +32,27 @@ Utility::canonicalizeHeaders(const Http::RequestHeaderMap& headers) {
         if (!entry.key().getStringView().empty() && entry.key().getStringView()[0] == ':') {
           return Http::HeaderMap::Iterate::Continue;
         }
+        // For S3, we restrict the headers even further to avoid  headers that might mutate
+        // when in transit (e.g.: x-forwarded-for, internal headers, etc).
+        if (*pair->second == "s3") {
```
Comment (marcomagdy) [inline source/extensions/common/aws/utility.cc]: This doesn't apply just to s3. You can remove this check.

File: source/extensions/common/aws/utility.cc [hunk 2]
```diff
@@ -25,19 +32,27 @@ Utility::canonicalizeHeaders(const Http::RequestHeaderMap& headers) {
...
+          const auto key = entry.key().getStringView();
+          if (key != Http::Headers::get().ContentType.get() && !absl::StartsWith(key, "x-amz-")) {
```
Comment (marcomagdy) [inline source/extensions/common/aws/utility.cc]: I'm not sure I understand those two checks.
Shouldn't it just be if `if (key != Http::Headers::get().ForwardedProto)` then continue?

File: source/extensions/common/aws/signer_impl.cc [hunk 3]
```diff
@@ -24,6 +24,9 @@ void SignerImpl::sign(Http::RequestMessage& message, bool sign_body) {
 }
 
 void SignerImpl::sign(Http::RequestHeaderMap& headers) {
+  // We are signing something without a body, so insert the empty string hash.
+  headers.addCopy(SignatureHeaders::get().ContentSha256,
+                  SignatureConstants::get().HashedEmptyString);
```
Comment (marcomagdy) [inline source/extensions/common/aws/signer_impl.cc]: This is one of the areas where s3 is different from all other aws services when it comes to signing. This needs to be the literal `UNSIGNED_PAYLOAD` (only for s3)

File: source/extensions/common/aws/utility.cc (hunk 2 above)
Comment (rgs1) [reply source/extensions/common/aws/utility.cc]: you mean `Http::Headers::get().ForwardedFor()`? `x-forwarded-proto` is less likely to change, no? 

More generally though, I want to reduce the set of headers that are signed and per https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html the bare minimum are: `host`, `content-type` and `x-amzn-*`.

If we allow a broader range of headers to be signed, users might get errors because of headers that are dropped/removed after hence causing the signature check to failure.

Or am I missing something?

File: source/extensions/common/aws/utility.cc (hunk 1 above)
Comment (rgs1) [reply source/extensions/common/aws/utility.cc]: True(-ish). Comparing https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html to https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html, it looks like S3 has a smaller set of minimally required headers -- no?

Also, S3 requests are more likely to see headers mutations given it's a typical use case to have regular HTTP traffic --> Envoy --> S3?

File: source/extensions/common/aws/signer_impl.cc (hunk 3 above)
Comment (rgs1) [reply source/extensions/common/aws/signer_impl.cc]: Note that the above works too.... But sure.

Comment (marcomagdy) [reply source/extensions/common/aws/signer_impl.cc]: It might work for GETs, but does it work for PUTs/uploading? I'd be surprised.

Comment (rgs1) [reply source/extensions/common/aws/signer_impl.cc]: i only tried GETs

**Comment by marcomagdy** [comment]
Spoke privately with @rgs1 and we decided to:
- Blacklist `x-forwarded-for` and `x-forwarded-proto` in the signer, since they do mutate in our experience. And it's better to sign as many headers as possible for security and for compatibility with all AWS services.
- Use `UNSIGNED_PAYLOAD` for S3 payloads only.
- `host_rewrite` is needed because in S3's case the bucket name is part of the domain.
- `prefix_rewrite` is needed for flexibility of URLs in general
- We need a `per_filter_config` for this filter.

**Comment by rgs1** [comment]
> Spoke privately with @rgs1 and we decided to:
> 
> * Blacklist `x-forwarded-for` and `x-forwarded-proto` in the signer, since they do mutate in our experience. And it's better to sign as many headers as possible for security and for compatibility with all AWS services.
> * Use `UNSIGNED_PAYLOAD` for S3 payloads only.
> * `host_rewrite` is needed because in S3's case the bucket name is part of the domain.
> * `prefix_rewrite` is needed for flexibility of URLs in general
> * We need a `per_filter_config` for this filter.

Thanks for the summary; handled all except:

1) `prefix_rewrite` (will push commits for that later this weekend or on Mon) 
2) `per_filter_config` support, since I want to tackle that via #10164

**Comment by rgs1** [comment]
> > Spoke privately with @rgs1 and we decided to:
> > 
> > * Blacklist `x-forwarded-for` and `x-forwarded-proto` in the signer, since they do mutate in our experience. And it's better to sign as many headers as possible for security and for compatibility with all AWS services.
> > * Use `UNSIGNED_PAYLOAD` for S3 payloads only.
> > * `host_rewrite` is needed because in S3's case the bucket name is part of the domain.
> > * `prefix_rewrite` is needed for flexibility of URLs in general
> > * We need a `per_filter_config` for this filter.
> 
> Thanks for the summary; handled all except:
> 
> 1. `prefix_rewrite` (will push commits for that later this weekend or on Mon)
> 2. `per_filter_config` support, since I want to tackle that via #10164

I don't actually need `prefix_rewrite` (which needs `per_filter_config`) immediately, so that can happen in a follow-up PR.

cc: @marcomagdy

File: source/extensions/common/aws/signer_impl.cc [hunk 4]
```diff
@@ -24,7 +24,16 @@ void SignerImpl::sign(Http::RequestMessage& message, bool sign_body) {
...
 
 void SignerImpl::sign(Http::RequestHeaderMap& headers) {
-  sign(headers, SignatureConstants::get().HashedEmptyString);
+  // S3 payloads require special treatment.
+  if (service_name_ == "s3") {
+    headers.addCopy(SignatureHeaders::get().ContentSha256,
+                    SignatureConstants::get().UnsignedPayload);
+    sign(headers, SignatureConstants::get().UnsignedPayload);
+  } else {
+    headers.addCopy(SignatureHeaders::get().ContentSha256,
```
Comment (marcomagdy) [inline source/extensions/common/aws/signer_impl.cc]: nit: I believe you can use setReference here instead of addCopy.

File: source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.h [hunk 5]
```diff
@@ -45,6 +45,11 @@ class FilterConfig {
    * @return the filter stats.
    */
   virtual FilterStats& stats() PURE;
+
+  /**
+   * @return the host rewrite value.
+   */
+  virtual const std::string& hostRewrite() PURE;
```
Comment (marcomagdy) [inline source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.h]: nit: method can be made const

File: test/extensions/common/aws/utility_test.cc [hunk 6]
```diff
@@ -84,6 +84,19 @@ TEST(UtilityTest, CanonicalizeHeadersTrimmingWhitespace) {
                           Pair("leading", "leading value"), Pair("trailing", "trailing value")));
 }
 
+// Headers that are likely to mutate are not considered canonical
+TEST(UtilityTest, CanonicalizeHeadersS3) {
```
Comment (marcomagdy) [inline test/extensions/common/aws/utility_test.cc]: nit: test name says s3, but this is unrelated to s3. rename?
